### PR TITLE
[Runtime Env] Make every ${ENV_VAR} replaceable

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

In the master, we only allow a few env vars specified via ${} translatable. This is unnecessarily too restrictive. This PR allows every env vars to be translated if they are specified via ${ENV_VAR}

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/utils.py [hunk 1]
```diff
@@ -1884,18 +1884,8 @@ def update_envs(env_vars: Dict[str, str]):
...
-        "DYLD_LIBRARY_PATH",
-        "LD_PRELOAD",
-    ]
-
     for key, value in env_vars.items():
-        if key in replaceable_keys:
-            os.environ[key] = value.replace("${" + key + "}", os.environ.get(key, ""))
-        else:
-            os.environ[key] = value
+        os.environ[key] = os.path.expandvars(value)
```
Comment (rkooo567) [inline python/ray/_private/utils.py:1888]: Looks like the Python implementation leaves the env var unchanged if it doesn't exist. E.g.,

```
unset ENV_VAR

# previous
${ENV_VAR} -> ""

# now
${ENV_VAR} -> ${ENV_VAR}
```

I feel like this should be fine (especially given this python implementation will work cross platform correctly). But if we have to keep this behavior, I can somehow write the custom implementation

**Review by jjyao — COMMENTED** [review]
Can you also update the doc (https://docs.ray.io/en/latest/ray-core/handling-dependencies.html) as well?

Currently it has
```
env_vars (Dict[str, str]): Environment variables to set. Environment variables already set on the cluster will still be visible to the Ray workers; so there is no need to include os.environ or similar in the env_vars field. By default, these environment variables override the same name environment variables on the cluster. You can also reference existing environment variables using ${ENV_VAR} to achieve the appending behavior. Only PATH, LD_LIBRARY_PATH, DYLD_LIBRARY_PATH, and LD_PRELOAD are supported. See below for an example:

Example: {"OMP_NUM_THREADS": "32", "TF_WARNINGS": "none"}

Example: {"LD_LIBRARY_PATH": "${LD_LIBRARY_PATH}:/home/admin/my_lib"}
```

File: python/ray/_private/utils.py (hunk 1 above)
Comment (jjyao) [reply python/ray/_private/utils.py:1888]: nice, using python's expandvars is fine

**Comment by rkooo567** [comment]
cc @jjyao it is ready for review.

**Review by jjyao — APPROVED** [review]
Do we have tests already?

**Comment by rkooo567** [comment]
@jjyao yes!

File: doc/source/ray-core/handling-dependencies.rst [hunk 2]
```diff
@@ -406,12 +406,14 @@ The ``runtime_env`` is a Python dictionary or a Python class :class:`ray.runtime
   no need to include ``os.environ`` or similar in the ``env_vars`` field.
   By default, these environment variables override the same name environment variables on the cluster.
   You can also reference existing environment variables using ${ENV_VAR} to achieve the appending behavior.
-  Only PATH, LD_LIBRARY_PATH, DYLD_LIBRARY_PATH, and LD_PRELOAD are supported. See below for an example:
+  If the environment variable doesn't exist, it will become an empty string `""`.
```
Comment (angelinalg) [inline doc/source/ray-core/handling-dependencies.rst:409]:
```suggestion
  If the environment variable doesn't exist, it becomes an empty string `""`.
```

File: doc/source/ray-core/handling-dependencies.rst [hunk 3]
```diff
@@ -406,12 +406,14 @@ The ``runtime_env`` is a Python dictionary or a Python class :class:`ray.runtime
...
 
   - Example: ``{"OMP_NUM_THREADS": "32", "TF_WARNINGS": "none"}``
 
   - Example: ``{"LD_LIBRARY_PATH": "${LD_LIBRARY_PATH}:/home/admin/my_lib"}``
 
+  - Non-existing variables Example: ``{"ENV_VAR_NOT_EXIST": "${ENV_VAR_NOT_EXIST}:/home/admin/my_lib"}`` -> ``ENV_VAR_NOT_EXIST=":/home/admin/my_lib"``.
```
Comment (angelinalg) [inline doc/source/ray-core/handling-dependencies.rst:415]:
```suggestion
  - Non-existant variable example: ``{"ENV_VAR_NOT_EXIST": "${ENV_VAR_NOT_EXIST}:/home/admin/my_lib"}`` -> ``ENV_VAR_NOT_EXIST=":/home/admin/my_lib"``.
```

File: python/ray/_private/utils.py [hunk 4]
```diff
@@ -2050,18 +2050,11 @@ def update_envs(env_vars: Dict[str, str]):
...
-        "LD_PRELOAD",
-    ]
-
     for key, value in env_vars.items():
-        if key in replaceable_keys:
-            os.environ[key] = value.replace("${" + key + "}", os.environ.get(key, ""))
-        else:
-            os.environ[key] = value
+        expanded = os.path.expandvars(value)
+        # Replace non-existing env vars to an empty string.
```
Comment (angelinalg) [inline python/ray/_private/utils.py:2055]:
```suggestion
        # Replace non-existant env vars with an empty string.
```

**Review by angelinalg — APPROVED** [review]

**Comment by rkooo567** [comment]
Merged it because the previous run succeeds, and I only made doc changes.

**Pull request merged**